utilities: Added script to bulk rename prefixes

garberg opened the pull request:
Added a script which can be used to "rename" prefixes in NIPAP, that is change description and node, in bulk. When executed, the script will display a list of the prefixes to change and what they will be changed to as well as let the user select what prefixes to perform the operation on.  Either all, a few selected prefixes or all prefixes EXCEPT the selected prefixes can be chosen.

plajjan commented:
I don't think "rename" is a good name for this script although I do understand where it originates from. Prefixes don't really have names.. this is more of pattern replacement, so can we rename the script accordingly?

Maybe one day we should consider integrating this in the CLI but I think scripts like this is good in the meantime.

garberg commented:
What about bulk-update?

plajjan commented on utilities/bulk-rename.py:
```diff
@@ -0,0 +1,179 @@
+    prefix nodes and descriptions, this tool replaces all instances of a string
+    with another.
+    
+    Currently the foolowing attributes are checked for strings to replace:
```
fool ;)

plajjan commented on utilities/bulk-rename.py:
```diff
@@ -0,0 +1,179 @@
+    prefix_list += prefix_result
+    print len(prefix_list), 
+    sys.stdout.flush()
+    while len(prefix_result) == 100:
```
I'm getting tired of this pattern. We should build a generator out of it so it's more natural to loop over prefixes.. It's not really related to your PR though ;)

garberg commented:
OK, fixed type and renamed the script. I noticed that I myself referred to the function of the script as string replacement, so I went with bulk-string-replace.py.

plajjan commented on utilities/bulk-string-replace.py:
```diff
@@ -0,0 +1,179 @@
+            if use_colors:
+                print "\x1b[0m",
+            print " -- %d --" % i
+            color = "\x1b[91m"
```
Does this really honour use_colors? You use it just above but here we set the color variable which is used on line 96 to print stuff. I don't know color codes by heart. Aren't there names for these in some lib or can we define them somewhere instead?

plajjan commented on utilities/bulk-string-replace.py:
```diff
@@ -0,0 +1,179 @@
+        if prefix.match:
+            new_prefix_node = re.sub(pattern, replacement, (prefix.node or ''), flags=re.IGNORECASE)
+            new_prefix_desc = re.sub(pattern, replacement, (prefix.description or ''), flags=re.IGNORECASE)
+            print ("\x1b[92m" + print_pattern) % (
```
Same thing about colours here, is use_color honoured or not?

plajjan commented on utilities/bulk-string-replace.py:
```diff
@@ -0,0 +1,179 @@
+    inp = raw_input("Selection: ").strip()
+
+    invert = False
+    if inp[0] == "!":
```
I guess this fails if inp is empty, i.e. user just pressed enter.

garberg replied:
Yup :smile:

plajjan commented on utilities/bulk-string-replace.py:
```diff
@@ -0,0 +1,179 @@
+    print "Select replacements to perform"
+    print "Enter comma-separated selection (eg. 5,7,10) or \"all\" for all prefixes."
+    print "Prefix list with ! to invert selection (eg !5,7,10 to perform operation on all except the entered prefixes)"
+    inp = raw_input("Selection: ").strip()
```
There doesn't appear to be a way to cancel any replacement!?

^C ?

Also, I think "inp" is a silly abbreviation of input ;)

garberg replied:
Yeah, ^C was the intention (even tough it throws a nasty traceback in the user's face).

I wanted to avoid input as that's the name of a built-in function, see https://docs.python.org/2/library/functions.html#input. I could probably have chosen a better name though.

plajjan commented on utilities/bulk-string-replace.py:
```diff
@@ -0,0 +1,179 @@
+    auth_uri = "%s:%s@" % (args.username or cfg.get('global', 'username'),
+            args.password or cfg.get('global', 'password'))
+
+    xmlrpc_uri = "http://%(auth_uri)s%(host)s:%(port)s" % {
```
guess this fails if I put in an IPv6 address here ;)

Most other scripts fail in a similar manner but perhaps we should start improving this. Can we always do [%(host)s]? I guess it works for IPv4 too, right?

garberg replied:
Yup, seems to work, I'll update. This was taken from boilerplate.py, also it should be updated.

garberg commented:
I think you vastly overestimate the ambition of this script :stuck_out_tongue: 

But sure, I'll do some updates. There seems to be a few libraries which assist with colors, for example https://pypi.python.org/pypi/colorama. Regarding use_colors, I guess that could be removed entirely. As it's an interactive script it'll probably be very rare that it'll write to anything other than a TTY.

plajjan commented:
@garberg haha, perhaps. I have no major issue merging it as is really.. it's better having the script in its current state than no script at all but just fixing the host thing for IPv6 and the empty input should be easy and then we're good to go.

I agree with you on the colors, I think you can assume color terminal. If that proves to be wrong then.. well, too bad :P

garberg commented:
Okay, some changes! Handle empty and invalid input, remove TTY verification, fixed issue with IPv6 hosts and replaced magic color characters with helpfully named variables. I chose not to use any terminal color library to avoid adding dependencies.
